### PR TITLE
[SPARK-45956][BUILD] Upgrade `Apache ZooKeeper` to 3.9.1

### PR DESCRIPTION
**Pull request by bjornjorgensen**
### What changes were proposed in this pull request?
Upgrade Apache ZooKeeper from 3.6.3 to 3.9.1
[Releasenotes 3.9.1](https://zookeeper.apache.org/doc/r3.9.1/releasenotes.html)

### Why are the changes needed?
[CVE-2023-44981](https://nvd.nist.gov/vuln/detail/CVE-2023-44981) [9.1 CRITICAL](https://nvd.nist.gov/vuln-metrics/cvss/v3-calculator?name=CVE-2023-44981&vector=AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:N&version=3.1&source=NIST)
[Explicit handling of DIGEST-MD5 vs GSSAPI in quorum auth](https://issues.apache.org/jira/browse/ZOOKEEPER-4753)

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by bjornjorgensen** [comment]
Link to runner https://github.com/bjornjorgensen/spark/actions/runs/6893856337

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for providing the GitHub Action link together, @bjornjorgensen .

BTW, we have Apache Zookeeper 3.9.1 already. Do you have any reason to choose 3.7.2? Is there any outage with 3.8.x or 3.9.x? Sorry for asking this question. I forgot the historic context why we stick to 3.6.x.
- https://github.com/apache/zookeeper/releases/tag/release-3.9.1

**Comment by dongjoon-hyun** [comment]
Yes, if there is no issue with Hadoop and Hive, it would be great for Apache Spark 4.0.0.

File: dev/deps/spark-deps-hadoop-3-hive-2.3 [hunk 1]
```diff
@@ -179,6 +179,8 @@ log4j-1.2-api/2.21.0//log4j-1.2-api-2.21.0.jar
 log4j-api/2.21.0//log4j-api-2.21.0.jar
 log4j-core/2.21.0//log4j-core-2.21.0.jar
 log4j-slf4j2-impl/2.21.0//log4j-slf4j2-impl-2.21.0.jar
+logback-classic/1.2.10//logback-classic-1.2.10.jar
+logback-core/1.2.10//logback-core-1.2.10.jar
```
Comment (dongjoon-hyun) [inline dev/deps/spark-deps-hadoop-3-hive-2.3:183]: Oh. Can we exclude this by reusing our existing log4j2 dependency?

Comment (bjornjorgensen) [reply dev/deps/spark-deps-hadoop-3-hive-2.3:183]: hmm.. zookeeper Migrate to Logback https://issues.apache.org/jira/browse/ZOOKEEPER-4427 in version 3.8.0 There are others that have asked the same question as you. https://github.com/apache/zookeeper/pull/1793#issuecomment-1557181105 but that will be a lot of work. Should we just go back to version 3.7.2 for this one?

Comment (dongjoon-hyun) [reply dev/deps/spark-deps-hadoop-3-hive-2.3]: - Thank you for the confirmation, @xhumanoid .
- Thank you for updating the PR, @bjornjorgensen . Please revisit the PR title and description according to the code change.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs).

cc @sunchao, too.

**Review by sunchao — APPROVED** [review]
LGTM too (pending CI)

**Comment by dongjoon-hyun** [comment]
Thank you all. Merged to master for Apache Spark 4.0.0.

**Comment by Dzeri96** [comment]
Please forgive my ignorance, but if a severe vulnerability is found in the dependencies, why doesn't the fix get merged into the currently supported version? 3.6.x -> 3.9.x should be an API-compatible upgrade in theory.

**Comment by bjornjorgensen** [comment]
Hi @Dzeri96 

It's a valid point, and I'd like to provide some context on why the fix for the vulnerability was applied to Spark 4.0.0 and not backported to the 3.x releases.
The decision is rooted in the Apache Spark community's release and maintenance policies, as well as the practical considerations of managing a large and active open-source project.

Apache Spark's Maintenance Policy: The Apache Spark project generally maintains feature release branches with bug fixes for a period of 18 months. While critical security fixes are a high priority, the process for backporting them to older release lines involves careful consideration by the community. It is not always guaranteed that dependency updates, even for security reasons, will be backported to all supported versions. The focus is often on ensuring the stability of maintenance releases.

Severity of the Vulnerability: The vulnerability in question, CVE-2023-44981, is indeed a critical issue in Apache ZooKeeper, with a CVSS score of 9.1. It allows for an authorization bypass if SASL Quorum Peer authentication is enabled. However, this is not a default configuration, (to my knowledge) which may have factored into the urgency and backporting decision within the Spark community. The recommended fix is to upgrade to ZooKeeper versions 3.9.1, 3.8.3, or 3.7.2.

Focus on Future Releases: The Apache Spark community, like many open-source projects, prioritizes the stability and security of the upcoming major and minor releases. The upgrade to ZooKeeper 3.9.1 was merged into the master branch for the upcoming Spark 4.0.0 release, ensuring that the latest version of Spark would not be affected.

API Compatibility and the Precautionary Principle: While the upgrade from ZooKeeper 3.6.x to 3.9.x is expected to be API-compatible for client libraries, any dependency change in a maintenance release of a project as large as Spark carries a risk of introducing unexpected regressions or breaking changes for some users' environments. The Spark community tends to be conservative about making non-essential changes to older, stable release branches to avoid any potential disruption.

Community Resources: Backporting and thoroughly testing a dependency upgrade across all supported 3.x versions of Spark requires significant effort from the community contributors and committers. These resources are often focused on developing and stabilizing the next major version.

**Comment by Dzeri96** [comment]
Hi @bjornjorgensen ,
First of all, thank you for the very comprehensive answer.

I have one thing to ask you though: where does the discussion leading to these decisions take place?

My company is doing a security audit for a certificate and any official discussion on why something is or isn't a threat would really help make our case in situations where we chose not to upgrade a package like Spark.

Additionally, knowing this information would help with threat modelling. In today's world, it's impossible to understand how every single (sub)dependency inside your project is used.

**Pull request closed without merging**